File: aat-android/src/main/java/ch/bailu/aat/services/editor/EditorInterface.java
package ch.bailu.aat.services.editor;

import ch.bailu.aat.gpx.GpxList;
import ch.bailu.aat.gpx.GpxPoint;
import ch.bailu.aat.gpx.GpxPointNode;
import ch.bailu.aat.gpx.interfaces.GpxType;
import ch.bailu.foc.Foc;

public interface EditorInterface {
    EditorInterface NULL = new EditorInterface() {

        @Override
        public void save() {}

        @Override
        public void setType(GpxType type) {}

        @Override
        public void remove() {}

        @Override
        public void add(GpxPoint point) {}

        @Override
        public void up() {}

        @Override
        public void down() {}

        @Override
        public boolean isModified() {
            return false;
        }

        @Override
        public GpxPointNode getSelected() {
            return null;
        }

        @Override
        public void select(GpxPointNode p) {}

        @Override
        public void saveTo(Foc path) {}

        @Override
        public void clear() {}

        @Override
        public void redo() {}

        @Override
        public void undo() {}

        @Override
        public void inverse() {}

        @Override
        public void attach(GpxList file) {}

        @Override
        public void fix() {}

        @Override
        public void simplify() {}

        @Override
        public void cutPreceding() {}

        @Override
        public void cutRemaining(){}

    };
    void save();
    void setType(GpxType type);
    void remove();
    void add(GpxPoint point);
    void up();
    void down();
    boolean isModified();

    GpxPointNode getSelected();
    void select(GpxPointNode p);

    void saveTo(Foc path);
    void clear();
    void redo();
    void undo();

    void inverse();
    void attach(GpxList toAttach);
    void fix();
    void simplify();
    void cutPreceding();
    void cutRemaining();
}

File: aat-lib/src/main/java/ch/bailu/aat_lib/service/editor/EditorInterface.kt
package ch.bailu.aat_lib.service.editor

import ch.bailu.aat_lib.gpx.GpxList
import ch.bailu.aat_lib.gpx.GpxPoint
import ch.bailu.aat_lib.gpx.GpxPointNode
import ch.bailu.aat_lib.gpx.interfaces.GpxType
import ch.bailu.foc.Foc

interface EditorInterface {
    fun save()
    fun unload()
    fun setType(type: GpxType)
    fun remove()
    fun add(point: GpxPoint)
    fun up()
    fun down()
    fun isModified(): Boolean

    fun getSelected(): GpxPointNode?
    fun select(p: GpxPointNode)

    fun saveTo(path: Foc)
    fun clear()
    fun redo()
    fun undo()

    fun inverse()
    fun attach(toAttach: GpxList)
    fun fix()
    fun simplify()
    fun cutPreceding()
    fun cutRemaining()

    companion object {
        val NULL: EditorInterface = object : EditorInterface {
            override fun save() {}

            override fun unload() {}

            override fun setType(type: GpxType) {}

            override fun remove() {}

            override fun add(point: GpxPoint) {}

            override fun up() {}

            override fun down() {}

            override fun isModified(): Boolean {
                return false
            }

            override fun getSelected(): GpxPointNode? {
                return null
            }

            override fun select(p: GpxPointNode) {}

            override fun saveTo(path: Foc) {}

            override fun clear() {}

            override fun redo() {}

            override fun undo() {}

            override fun inverse() {}

            override fun attach(file: GpxList) {}

            override fun fix() {}

            override fun simplify() {}

            override fun cutPreceding() {}

            override fun cutRemaining() {}
        }
    }
}
